pkg/internal/client/txn.go

```diff
@@ -572,9 +572,38 @@ func (txn *Txn) AddCommitTrigger(trigger func()) {
 	txn.commitTriggers = append(txn.commitTriggers, trigger)
 }
 
+// maybeFinishReadonly provides a fast-path for finishing a read-only
+// transaction without going through the overhead of creating an
+// EndTransactionRequest only to not send it.
+//
+// NB: The logic here must be kept in sync with the logic in txn.Send.
+func (txn *Txn) maybeFinishReadonly(commit bool, deadline *hlc.Timestamp) (bool, *roachpb.Error) {
+	txn.mu.Lock()
+	defer txn.mu.Unlock()
+	if txn.mu.Proto.Writing || txn.mu.writingTxnRecord {
+		return false, nil
+	}
+	txn.mu.finalized = true
+	// Check that read only transactions do not violate their deadline. This can NOT
+	// happen since the txn deadline is normally updated when it is about to expire
+	// or expired. We will just keep the code for safety (see TestReacquireLeaseOnRestart).
+	if deadline != nil && deadline.Less(txn.mu.Proto.Timestamp) {
+		return false, roachpb.NewErrorWithTxn(roachpb.NewTransactionAbortedError(), &txn.mu.Proto)
```

Comment:
Good question. The code in `txn.Send` does exactly the same thing. I think it is safe because there will be no more updates to `txn.mu.Proto` because we set `txn.mu.finalized`.